indexing-service/src/main/java/org/apache/druid/indexing/input/DruidSegmentInputFormat.java

```diff
@@ -55,14 +56,30 @@ public InputEntityReader createReader(
       File temporaryDirectory
   )
   {
-    return new DruidSegmentReader(
-        source,
-        indexIO,
-        inputRowSchema.getTimestampSpec(),
-        inputRowSchema.getDimensionsSpec(),
-        inputRowSchema.getColumnsFilter(),
-        dimFilter,
-        temporaryDirectory
+    // this method handles the case when the entity comes from a tombstone or from a regular segment
+    Preconditions.checkArgument(
+        source instanceof DruidSegmentInputEntity,
+        DruidSegmentInputEntity.class.getName() + " required, but "
+        + source.getClass().getName() + " provided."
     );
+
+    InputEntityReader retVal = null;
+    // source is tombstone case
+    if ((source instanceof DruidSegmentInputEntity) && ((DruidSegmentInputEntity) source).isFromTombstone()) {
```

Comment:
nit: you are basically doing this same instanceof twice now (one for the precondition and one here, might as well just make the pre-condition an `if` statement and then cast and store a method-local variable.

Reply:
yeah.. the second check is redundant, by the time the code gets there the cast should always work due to the precondition passing...